### ISSUE TITLE
crash on deeply nested parentheses in input

### ISSUE THREAD

**Issue by jciena** [CastXML/CastXML#216]
I am trying to run castxml with the following options:
_castxml.exe --castxml-gccxml -o output.xml --castxml-cc-msvc-c "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.33.31629\bin\Hostx64\x64\cl.exe" xbd_win.h_

But getting the error:
Stack dump:
0. Program arguments: ..\\..\\..\\gccxml\\bin\\castxml.exe --castxml-gccxml -o output.xml --castxml-cc-msvc-c "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.33.31629\\bin\\Hostx64\\x64\\cl.exe" xbd_win.h
1 xbd_win.h:1168:42 <Spelling=xbd_win.h:343:46>: **current parser token ')'**

xbd_win.h, line 343
```
337 #define GROSSLOSSSEVPCT    274
(..)
343 #define TBASEFRAC_OFF      (GROSSLOSSSEVPCT+1)
```

And 1168
```
1168 TS_BondVariable_EURSWAP5Y =          EURSWAP5Y
```

EURSWAP5Y expands to (EURSWAP4Y+1)

My  castxml version is 0.4.5:
_CastXML project maintained and supported by Kitware (kitware.com).
clang version 13.0.0 (https://github.com/CastXML/CastXMLSuperbuild 3a0a2c545e708f63908639f7ea8799b5f03b8fc0)
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir:_

Am I using the right options, is there something missing?

**Comment by bradking** [comment]
Please strip `xbd_win.h` down to a minimal complete example input file that demonstrates the problem.

**Comment by bradking** [comment]
I can reproduce the problem with a simple `.c` source with deeply nested parentheses:

```c
enum E { X = ((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+
1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1) };
```

**Comment by jciena** [comment]
I do not see `-fbracket-depth` option in help. As you mentioned, in windows it makes no difference using it or not.
Should I report to llvm as asked in the error message below?

```
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: castxml.exe --castxml-gccxml -o output.xml --castxml-cc-msvc-c cl .\\input.c -fbracket-depth=1000
1.      .\input.c:1:194: current parser token '('

PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: castxml.exe --castxml-gccxml -o output.xml --castxml-cc-msvc-c cl .\\input.c
1.      .\input.c:1:194: current parser token '('
```

**Comment by bradking** [comment]
We need to try building CastXML with the latest LLVM version before reporting it there.

**Comment by jciena** [comment]
I already raised it [(link) there](https://github.com/llvm/llvm-project/issues/57456), and they suggested:

> Could you please try 15 or main branch? https://godbolt.org/ should be helpful.

I will build using the main and test again.

**Comment by bradking** [comment]
Thanks.  I think CastXML may need some updates to compile against LLVM/Clang 14 and/or 15.  I'll look at that.

**Comment by jciena** [comment]
I appreciate your support.

**Comment by bradking** [comment]
CastXML builds cleanly against LLVM/Clang 14.0.  However, the LLVM/Clang `release/15.x` branch currently has a regression that breaks CastXML.  I've reported that as llvm/llvm-project#57483.

**Comment by bradking** [comment]
After hacking the LLVM/Clang 15.x install tree with this patch:

<details>

```patch
--- include/clang/Lex/Preprocessor.h.orig       2022-08-31 09:24:20.359922800 -0400
+++ include/clang/Lex/Preprocessor.h    2022-08-31 11:55:51.490578400 -0400
@@ -1353,6 +1353,8 @@
   StringRef getLastMacroWithSpelling(SourceLocation Loc,
                                      ArrayRef<TokenValue> Tokens) const;

+  const std::string &getPredefines() const { return Predefines; }
+
   /// Set the predefines for this Preprocessor.
   ///
   /// These predefines are automatically injected when parsing the main file.
```

</details>

then CastXML builds again and all tests pass.

However, the crash discussed here remains.  I've got a Debug build, so I'll try to get a backtrace.

**Comment by bradking** [comment]
Clang's internal recursion overflows the stack before reaching the default `-fbracket-depth=256` nesting level limit.  I've posted more details in https://github.com/llvm/llvm-project/issues/57456#issuecomment-1233159203.

**Comment by bradking** [comment]
The problem doesn't occur in LLVM/Clang's own binaries because they are built with a larger stack: https://github.com/llvm/llvm-project/issues/57456#issuecomment-1233170017.

CastXML should be built with a larger stack too.

**Comment by bradking** [comment]
I've released [CastXML 0.4.6](https://github.com/CastXML/CastXML/releases/tag/v0.4.6) with this fix.

**Comment by jciena** [comment]
Thank you, I'll check it.
Much appreciated!
